src/Fixer/PhpTag/ShortEchoTagFixer.php

```diff
@@ -0,0 +1,242 @@
+            ],
+            null
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        if (self::FORMAT_SHORT === $this->configuration[self::OPTION_FORMAT]) {
+            if (null !== $tokens->findSequence([[\T_OPEN_TAG], [\T_ECHO]])) {
+                return true;
+            }
+            if (null !== $tokens->findSequence([[\T_OPEN_TAG], [\T_PRINT]])) {
+                return true;
+            }
+
+            return false;
+        }
+
+        return $tokens->isTokenKindFound(\T_OPEN_TAG_WITH_ECHO);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition()
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder(self::OPTION_FORMAT, 'How the fixer should process short/long echo tags'))
+                ->setAllowedValues($this->supportedFormatOptions)
+                ->setDefault(self::FORMAT_LONG)
+                ->getOption(),
+            (new FixerOptionBuilder(self::OPTION_LONG_FUNCTION, 'The function to be used to expand the short echo tags'))
+                ->setAllowedValues($this->supportedLongFunctionOptions)
+                ->setDefault(self::LONG_FUNCTION_ECHO)
+                ->getOption(),
+            (new FixerOptionBuilder(self::OPTION_SHORT_ALWAYS, 'Always render short-echo tags even in case of complex code'))
+                ->setAllowedTypes(['bool'])
+                ->setDefault(false)
+                ->getOption(),
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(SplFileInfo $file, Tokens $tokens)
+    {
+        if (self::FORMAT_SHORT === $this->configuration[self::OPTION_FORMAT]) {
+            $this->longToShort($file, $tokens);
+        } else {
+            $this->shortToLong($file, $tokens);
+        }
+    }
+
+    private function longToShort(SplFileInfo $file, Tokens $tokens)
+    {
+        $offset = 0;
+        for (;;) {
+            $foundEcho = $tokens->findSequence([[\T_OPEN_TAG], [\T_ECHO]], $offset);
+            $foundPrint = $tokens->findSequence([[\T_OPEN_TAG], [\T_PRINT]], $offset);
+            if (null === $foundEcho) {
+                $found = $foundPrint;
+                if (null === $found) {
+                    break;
+                }
+            } elseif (null === $foundPrint || key($foundEcho) < key($foundPrint)) {
+                $found = $foundEcho;
+            } else {
+                $found = $foundPrint;
+            }
+            $startRange = key($found);
+            $offset = $startRange + 1;
+            if (!$this->configuration[self::OPTION_SHORT_ALWAYS] && $this->isComplexCode($tokens, $offset)) {
+                continue;
+            }
+            next($found);
+            $endRange = key($found);
+            $newTokens = $this->buildLongToShortTokens($tokens, $startRange, $endRange);
```

Comment:
Can't we simply remove the `echo`/`print` token with the following whitespace - if any - and override the tag open?

Reply:
It's not that simple. For example:
```
<? /* comment */ echo 1 ?>
```
should become
```
<?= /* comment */ 1 ?>
```

(it's included in the fixer tests).

Reply:
I don't understand - that's exactly what I'm suggesting - `echo`  and ` ` after it are removed and `<?' becomes `<?=`.

Reply:
Sorry, I didn't understand what you meant.

About removing the whitespace, we have two options: remove the whitespaces after the PHP opening tag, or remove the whitespaces after the `echo`/`print`.

Consider these two cases:

```
<?php /*comment*/ echo
  1
?>

<?php
  /*comment*/ echo 1
?>
```

what should be the short-echo version of them?